### PR TITLE
Add first class gc setting

### PR DESCRIPTION
**Pull request by robert3005**
with sane defaults

cc @schlosna @iamdanfox

### CODE REVIEW

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.groovy [hunk 1]
```diff
@@ -59,16 +59,29 @@ class JavaServiceDistributionPlugin implements Plugin<Project> {
 
         CreateStartScripts startScripts = CreateStartScriptsTask.createStartScriptsTask(project, 'createStartScripts')
         project.afterEvaluate {
-            CreateStartScriptsTask.configure(startScripts, distributionExtension.mainClass, distributionExtension.serviceName,
-                    distributionExtension.defaultJvmOpts, distributionExtension.enableManifestClasspath)
+            def jvmOpts = distributionExtension.defaultJvmOpts + (distributionExtension.heapSize != null
+                    ? ["-Xmx" + distributionExtension.heapSize, "-Xms" + distributionExtension.heapSize] : [])
```
Comment (markelliot) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.groovy]: We usually want to pick these up from our config management system/would expect these to be parameterized by map keys. Maybe consider pushing further upstream into named parameters in go-java-launcher?

Comment (robert3005) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.groovy]: Makes sense - will remove from this pr and figure out correct wiring

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LaunchConfigTask.groovy [hunk 2]
```diff
@@ -31,6 +31,22 @@ class LaunchConfigTask extends DefaultTask {
 
     static final List<String> tmpdirJvmOpts = ['-Djava.io.tmpdir=var/data/tmp']
 
+    static final Map<String, List<String>> gcProfiles = [
+            'throughput'   : [],
```
Comment (robert3005) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LaunchConfigTask.groovy]: This means we would use jvm default which switched to G1 in 9. Need to put correct settings but at the same time java releases claim that from 10 onwards G1 should supersede parallelold collector

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LaunchConfigTask.groovy [hunk 3]
```diff
@@ -21,16 +21,35 @@ import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
...
 class LaunchConfigTask extends DefaultTask {
 
     static final List<String> tmpdirJvmOpts = ['-Djava.io.tmpdir=var/data/tmp']
 
+    static final Map<String, List<String>> gcProfiles = [
+            'throughput'   : ['-XX:+UseParallelOldGC'],
+            'response-time': ['-XX:+UseParNewGC',
+                              '-XX:+UseConcMarkSweepGC',
+                              '-XX:+CMSParallelInitialMarkEnabled',
+                              '-XX:+CMSParallelRemarkEnabled',
```
Comment (schlosna) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LaunchConfigTask.groovy]: We probably don't need to set `-XX:+CMSParallelInitialMarkEnabled -XX:+CMSParallelRemarkEnabled` as those should default to true with Java 8+, see https://github.com/AdoptOpenJDK/openjdk-jdk8u/blame/fcb5661ae078d0e3521a1237a67b3039b1bfdbe3/hotspot/src/share/vm/runtime/globals.hpp#L1792

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LaunchConfigTask.groovy [hunk 4]
```diff
@@ -21,16 +21,35 @@ import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
...
+                              '-XX:+UseCMSInitiatingOccupancyOnly',
+                              '-XX:CMSInitiatingOccupancyFraction=68',
+                              '-XX:+CMSClassUnloadingEnabled'],
```
Comment (schlosna) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LaunchConfigTask.groovy]: We probably want to add the following as well for CMS:

```
-XX:+ExplicitGCInvokesConcurrent
-XX:+ClassUnloadingWithConcurrentMark
```

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LaunchConfigTask.groovy [hunk 5]
```diff
@@ -21,16 +21,35 @@ import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
...
     static final List<String> tmpdirJvmOpts = ['-Djava.io.tmpdir=var/data/tmp']
 
+    static final Map<String, List<String>> gcProfiles = [
+            'throughput'   : ['-XX:+UseParallelOldGC'],
+            'response-time': ['-XX:+UseParNewGC',
+                              '-XX:+UseConcMarkSweepGC',
+                              '-XX:+CMSParallelInitialMarkEnabled',
+                              '-XX:+CMSParallelRemarkEnabled',
+                              '-XX:+UseCMSInitiatingOccupancyOnly',
+                              '-XX:CMSInitiatingOccupancyFraction=68',
```
Comment (schlosna) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LaunchConfigTask.groovy]: would love a way to make this percentage configurable as different workloads may need adjusted

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionExtension.groovy [hunk 6]
```diff
@@ -21,6 +21,7 @@ import org.gradle.api.Project
 class JavaServiceDistributionExtension extends BaseDistributionExtension {
 
     private String mainClass
+    private String gcProfile = "throughput"
```
Comment (schlosna) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionExtension.groovy]: do we actually want `throughput` as default?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LaunchConfigTask.groovy (hunk 5 above)
Comment (robert3005) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LaunchConfigTask.groovy]: so this still can be overridden by anything in launcher-custom. Not sure what's the best api here. I have never seen it set to anything else internally

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionExtension.groovy (hunk 6 above)
Comment (robert3005) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionExtension.groovy]: really - I just went with what's the current setup. For java 8 we might want to go with response-time.

**Comment by robert3005** [comment]
Have a nicer way of expressing this - will update the pr tomorrow

**Comment by robert3005** [comment]
updated with new approach - should let us be specific while only exposing the things we think ought to be exposed

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionExtension.groovy [hunk 7]
```diff
@@ -117,4 +140,19 @@ class JavaServiceDistributionExtension extends BaseDistributionExtension {
     List<String> getExcludeFromVar() {
         return excludeFromVar
     }
+
+    void gc(String type, @Nullable Closure configuration) {
+        gc = objects.newInstance(profileNames[type])
```
Comment (robert3005) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionExtension.groovy:145]: It looks like `plugins { ... } ` block doesn't add the plugin classpath to buildscript classpath so we can't reference classes directly, i.e. doing

```
gc (ResponseTime) { ... }
```

fails with MissingPropertyException `ResponseTime`. We could require use of classpath but that seemed a bit suboptimal. Happy to change this if there are better options.

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionExtension.groovy [hunk 8]
```diff
@@ -117,4 +140,19 @@ class JavaServiceDistributionExtension extends BaseDistributionExtension {
...
+        if (configuration != null) {
```
Comment (robert3005) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionExtension.groovy:146]: need to add test that makes sure we can just do `gc 'hybrid'` for instance

Comment (robert3005) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionExtension.groovy:146]: added

**Review by ferozco — APPROVED** [review]
Since the default behaviour for java8 is the same as the previous behaviour, i think thats its good to go

**Pull request merged**